app/models/category_user.rb

```diff
@@ -23,20 +19,31 @@ def self.watching_levels
   end
 
   def self.batch_set(user, level, category_ids)
-    records = CategoryUser.where(user: user, notification_level: notification_levels[level])
-    old_ids = records.pluck(:category_id)
-    changed = false
-
-    category_ids = Category.where('id in (?)', category_ids).pluck(:id)
-
-    remove = (old_ids - category_ids)
-    if remove.present?
-      records.where('category_id in (?)', remove).destroy_all
-      changed = true
-    end
-
-    (category_ids - old_ids).each do |id|
-      CategoryUser.create!(user: user, category_id: id, notification_level: notification_levels[level])
+    level_num = notification_levels[level]
+    category_ids = Category.where(id: category_ids).pluck(:id)
+
+    # Update pre-existing category users
+    changed =
+      CategoryUser
+        .where(user_id: user.id, category_id: category_ids)
+        .update_all(notification_level: level_num) > 0
+
+    # Remove extraneous category users
+    changed ||=
+      CategoryUser
+        .where(user_id: user.id, notification_level: level_num)
+        .where.not(category_id: category_ids)
+        .delete_all() > 0
+
+    # Create new category users
+    new_category_ids =
```

Comment:
It would probably be better to use SQL here. You could write one insert statement for all the records, and because of the unique index do a `ON CONFLICT DO NOTHING`.

It's faster, but also safer under concurrency.

Reply:
@eviltrout, I've done this. I used the `connection.exec_query` instead of `DB.exec` because I didn't want to use a different db connection as that would have complicated the tests.

Rails 6 has bulk inserts, [see here](https://blog.bigbinary.com/2019/04/15/bulk-insert-support-in-rails-6.html), so I hope this code doesn't last too long.